### DISCUSSION TITLE
Old-style __getitem__ iterables are not accepted by `collections.abc.Iterable`

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#6223] [Q&A]
`Dataset`s in PyTorch only require you to define `__getitem__` and `__len__`. Since it doesn't define `__iter__`, this type of object is iterable as described in [PEP 234](https://peps.python.org/pep-0234/). `pyright` seems to be aware of this (although IMO it misses an opportunity to catch an error in the case that `__getitem__` is not hinted as accepting an `int`):

```python
from collections.abc import Mapping, Sequence
from typing import Generic, Protocol, TypeVar

T = TypeVar("T", covariant=True)

class MapStyleDataset(Generic[T], Protocol):
    def __len__(self) -> int:
        ...

    def __getitem__(self, index: str) -> T:
        ...

class MyDataset(MapStyleDataset[T]):
    def __init__(self, data: Mapping[str, T]) -> None:
        super().__init__()
        self.data = data

    def __len__(self) -> int:
        return len(self.data)
    
    def __getitem__(self, index: str) -> T:
        return self.data[index]

for el in MyDataset({"a": 1, "b": 2}):  # KeyError: 0, no pyright errors
    print(el)
```

So... my problem is that things typed as accepting `Iterable` (like `enumerate`) are all specifically looking for `__iter__` and don't account for this type of iterable object:

```python
from collections.abc import Sequence
from typing import Generic, Protocol, TypeVar

T = TypeVar("T", covariant=True)

class MapStyleDataset(Generic[T], Protocol):
    def __len__(self) -> int:
        ...

    def __getitem__(self, index: int) -> T:
        ...

class MyDataset(MapStyleDataset[T]):
    def __init__(self, data: Sequence[T]) -> None:
        super().__init__()
        self.data = data

    def __len__(self) -> int:
        return len(self.data)
    
    def __getitem__(self, index: int) -> T:
        return self.data[index]

for i, el in enumerate(MyDataset([0, 1, 2])):  # "MyDataset[int]" is incompatible with protocol "Iterable[_T@enumerate]"; "__iter__" is not present
    print(i, el)
# 0 0
# 1 1
# 2 2

```

I wonder if a type accepting either style of iterable couldn't be added to `collections.abc`? It could be a `TypeAlias` that is the union of `collections.abc.Iterable` and a `Protocol` like:

```python
class SequenceStyleIterable(Generic[T], Protocol):
    def __getitem__(self, index: int) -> T:
        ...
```

Or could it be worth hacking `pyright` accept these sequence-style iterables as `Iterable`? I am curious what you think the right solution could be.

**Comment by erictraut** [comment]
See [this issue](https://github.com/microsoft/pyright/issues/5433) for a discussion of this topic.
